File: src/skillguard/reporting/sarif_report.py
"""SARIF (Static Analysis Results Interchange Format) report generator.

Generates SARIF v2.1.0 compatible output for integration with GitHub
Code Scanning, VS Code, and other SARIF-consuming tools.
"""

from __future__ import annotations

import json
from typing import Any

from skillguard.core.models import ScanResult, Severity

# SARIF severity levels mapping
_SEVERITY_MAP: dict[str, str] = {
    "critical": "error",
    "high": "error",
    "medium": "warning",
    "low": "note",
    "info": "note",
}


def generate_sarif_report(result: ScanResult) -> str:
    """Generate a SARIF 2.1.0 report from scan results.

    Returns:
        JSON string in SARIF format.
    """
    # Collect all unique rules
    rules_map: dict[str, dict[str, Any]] = {}
    results_list: list[dict[str, Any]] = []

    for engine_result in result.engine_results:
        for finding in engine_result.findings:
            # Build rule entry
            if finding.rule_id not in rules_map:
                rule_entry: dict[str, Any] = {
                    "id": finding.rule_id,
                    "name": finding.rule_name,
                    "shortDescription": {"text": finding.rule_name},
                    "fullDescription": {"text": finding.description},
                    "defaultConfiguration": {
                        "level": _SEVERITY_MAP.get(finding.severity.value, "warning")
                    },
                    "properties": {
                        "tags": [finding.category],
                    },
                }
                if finding.owasp_llm:
                    rule_entry["properties"]["owasp_llm"] = finding.owasp_llm
                if finding.mitre_attack:
                    rule_entry["properties"]["mitre_attack"] = finding.mitre_attack
                if finding.cwe:
                    rule_entry["properties"]["cwe"] = finding.cwe
                rules_map[finding.rule_id] = rule_entry

            # Build result entry
            sarif_result: dict[str, Any] = {
                "ruleId": finding.rule_id,
                "level": _SEVERITY_MAP.get(finding.severity.value, "warning"),
                "message": {"text": finding.description},
                "locations": [
                    {
                        "physicalLocation": {
                            "artifactLocation": {"uri": finding.file_path},
                            "region": {},
                        }
                    }
                ],
            }

            if finding.line_start is not None:
                sarif_result["locations"][0]["physicalLocation"]["region"][
                    "startLine"
                ] = finding.line_start
            if finding.line_end is not None:
                sarif_result["locations"][0]["physicalLocation"]["region"][
                    "endLine"
                ] = finding.line_end
            if finding.snippet:
                sarif_result["locations"][0]["physicalLocation"]["region"][
                    "snippet"
                ] = {"text": finding.snippet}

            if finding.remediation:
                sarif_result["fixes"] = [
                    {
                        "description": {"text": finding.remediation},
                    }
                ]

            results_list.append(sarif_result)

    # Build SARIF document
    sarif: dict[str, Any] = {
        "$schema": "https://raw.githubusercontent.com/oasis-tcs/sarif-spec/main/sarif-2.1/schema/sarif-schema-2.1.0.json",
        "version": "2.1.0",
        "runs": [
            {
                "tool": {
                    "driver": {
                        "name": "SkillGuard",
                        "version": "0.1.0",
                        "informationUri": "https://github.com/skillguard/skillguard",
                        "rules": list(rules_map.values()),
                    }
                },
                "results": results_list,
                "properties": {
                    "skillName": result.skill_name,
                    "skillSha256": result.skill_sha256,
                    "compositeScore": result.composite_score,
                    "verdict": result.verdict if isinstance(result.verdict, str) else result.verdict.value,
                    "filesScanned": result.files_scanned,
                },
            }
        ],
    }

    return json.dumps(sarif, indent=2)

File: go-skillguard/internal/reporting/sarif_report.go
package reporting

import (
	"encoding/json"

	"github.com/girdav01/skillguard/internal/core"
)

var sarifSeverityMap = map[core.Severity]string{
	core.SeverityCritical: "error",
	core.SeverityHigh:     "error",
	core.SeverityMedium:   "warning",
	core.SeverityLow:      "note",
	core.SeverityInfo:     "note",
}

// GenerateSARIFReport generates a SARIF 2.1.0 report.
func GenerateSARIFReport(result *core.ScanResult) (string, error) {
	rulesMap := make(map[string]map[string]any)
	var resultsList []map[string]any

	for _, er := range result.EngineResults {
		for _, f := range er.Findings {
			if _, exists := rulesMap[f.RuleID]; !exists {
				rule := map[string]any{
					"id":               f.RuleID,
					"name":             f.RuleName,
					"shortDescription": map[string]string{"text": f.RuleName},
					"fullDescription":  map[string]string{"text": f.Description},
					"defaultConfiguration": map[string]string{
						"level": sarifSeverityMap[f.Severity],
					},
					"properties": map[string]any{
						"tags": []string{f.Category},
					},
				}
				if len(f.OWASPLLM) > 0 {
					rule["properties"].(map[string]any)["owasp_llm"] = f.OWASPLLM
				}
				if len(f.MITREAttack) > 0 {
					rule["properties"].(map[string]any)["mitre_attack"] = f.MITREAttack
				}
				if f.CWE != nil {
					rule["properties"].(map[string]any)["cwe"] = *f.CWE
				}
				rulesMap[f.RuleID] = rule
			}

			region := map[string]any{}
			if f.LineStart != nil {
				region["startLine"] = *f.LineStart
			}
			if f.LineEnd != nil {
				region["endLine"] = *f.LineEnd
			}
			if f.Snippet != nil {
				region["snippet"] = map[string]string{"text": *f.Snippet}
			}

			sarifResult := map[string]any{
				"ruleId":  f.RuleID,
				"level":   sarifSeverityMap[f.Severity],
				"message": map[string]string{"text": f.Description},
				"locations": []map[string]any{
					{
						"physicalLocation": map[string]any{
							"artifactLocation": map[string]string{"uri": f.FilePath},
							"region":           region,
						},
					},
				},
			}

			if f.Remediation != nil {
				sarifResult["fixes"] = []map[string]any{
					{"description": map[string]string{"text": *f.Remediation}},
				}
			}

			resultsList = append(resultsList, sarifResult)
		}
	}

	rules := make([]map[string]any, 0, len(rulesMap))
	for _, r := range rulesMap {
		rules = append(rules, r)
	}

	verdictStr := string(result.Verdict)

	sarif := map[string]any{
		"$schema": "https://raw.githubusercontent.com/oasis-tcs/sarif-spec/main/sarif-2.1/schema/sarif-schema-2.1.0.json",
		"version": "2.1.0",
		"runs": []map[string]any{
			{
				"tool": map[string]any{
					"driver": map[string]any{
						"name":           "SkillGuard",
						"version":        "0.1.0",
						"informationUri": "https://github.com/skillguard/skillguard",
						"rules":          rules,
					},
				},
				"results": resultsList,
				"properties": map[string]any{
					"skillName":      result.SkillName,
					"skillSha256":    result.SkillSHA256,
					"compositeScore": result.CompositeScore,
					"verdict":        verdictStr,
					"filesScanned":   result.FilesScanned,
				},
			},
		},
	}

	data, err := json.MarshalIndent(sarif, "", "  ")
	if err != nil {
		return "", err
	}
	return string(data), nil
}
